resteasy-core/src/main/java/org/jboss/resteasy/plugins/providers/MultiValuedParamConverterProvider.java

```diff
@@ -0,0 +1,173 @@
+package org.jboss.resteasy.plugins.providers;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.regex.Pattern;
+
+import javax.ws.rs.CookieParam;
+import javax.ws.rs.HeaderParam;
+import javax.ws.rs.ext.ParamConverter;
+import javax.ws.rs.ext.ParamConverterProvider;
+import javax.ws.rs.ext.Provider;
+
+import org.jboss.resteasy.annotations.Separator;
+import org.jboss.resteasy.core.ResteasyContext;
+import org.jboss.resteasy.core.StringParameterInjector;
+import org.jboss.resteasy.resteasy_jaxrs.i18n.LogMessages;
+import org.jboss.resteasy.spi.ResteasyProviderFactory;
+import org.jboss.resteasy.spi.util.Types;
+
+/**
+ * @author Marek Kopecky mkopecky@redhat.com
+ * @author Ron Sigal rsigal@redhat.com
+ */
+@Provider
+public class MultiValuedParamConverterProvider implements ParamConverterProvider
+{
+   private static final Pattern PROXY_REGEX = Pattern.compile("\\p{Punct}|\\[\\p{Punct}+\\]");
+
+   @SuppressWarnings("unchecked")
+   @Override
+   public <T> ParamConverter<T> getConverter(Class<T> rawType, Type genericType, Annotation[] annotations)
+   {
+      if (!isApplicable(annotations))
+      {
+         return null;
+      }
+      String separator = getSeparator(annotations);
+      Class<?> paramType = getHeaderParam(annotations);
```

Comment:
The `paramType` here should never be null to avoid NPE in `StringParameterInjector.getParamSignature()` wich is used many times in `StringParameterInjector` class.
So I suggest to change and rename `getHeaderParam()` this way:

```
 private Class<?> getParamType(Annotation[] annotations)
   {
      if (annotations == null)
      {
         return null;
      }
      for (Annotation a : annotations)
      {
         if (a instanceof HeaderParam)
         {
            return HeaderParam.class;
         }
         else if (a instanceof CookieParam)
         {
            return CookieParam.class;
         }
         else if (a instanceof PathParam)
         {
            return PathParam.class;
         }
         else if (a instanceof MatrixParam)
         {
            return MatrixParam.class;
         }
         else if (a instanceof QueryParam)
         {
            return QueryParam.class;
         }
         else if (a instanceof FormParam)
         {
            return FormParam.class;
         }
      }
      return null;
   }
```

And then if we go further, this `MultiValuedParamConverterProvider` is applicable if the `@Separator` is present and the `paramType` is not null so the code can be changed to:

```
@Provider
public class MultiValuedParamConverterProvider implements ParamConverterProvider
{

@SuppressWarnings("unchecked")
@Override
public <T> ParamConverter<T> getConverter(Class<T> rawType, Type genericType, Annotation[] annotations)
{
  Separator separatorAnnotation = getSeparatorAnnotation(annotations);
  Class<?> paramType = getParamType(annotations);
  if(separatorAnnotation == null || paramType == null){
	 return null;
  }
  String separator = getSeparator(separatorAnnotation,paramType);
  ...
}

...

private Separator getSeparatorAnnotation(Annotation[] annotations)
{
  for (Annotation a : annotations)
  {
	 if (a instanceof Separator)
	 {
		return (Separator) a;
	 }
  }
  return null;
}

private String getSeparator(Separator separatorAnnoation, Class<?> paramType)
{
  String separator = separatorAnnoation.value();
  if ("".equals(separator))
  {
	 separator = CookieParam.class.equals(paramType) ? "-" : ",";
  }
  return separator;
}

...
}
```

WDYT?

Reply:
@NicoNes, well, I care only about @HeaderParam. But I've changed getParamSignature() to
```
   public String getParamSignature()
   {
      return (paramType != null ? paramType.getName() : "") + "(\"" + paramName + "\")";
   }
```
Also, now that I think of it, I changed
```
       if (HeaderParam.class.equals(paramType))
       {
          delegate = factory.getHeaderDelegate(baseType);
          if (delegate != null)
          {
        	  return true;
          }
       }
```
to
```
       if (HeaderParam.class.equals(paramType) || org.jboss.resteasy.annotations.jaxrs.HeaderParam.class.equals(paramType))
       {
          delegate = factory.getHeaderDelegate(baseType);
          if (delegate != null)
          {
        	  return true;
          }
       }
```

Reply:
Ok :+1: